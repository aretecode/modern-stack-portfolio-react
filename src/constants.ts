export const defaultApolloStatePortfolio = {
  __typename: 'Resume',
  basics: {
    __typename: 'Basics',
    name: '',
    label: '',
    picture: '',
    email: 'james@jameswiens.com',
    telephone: '12506509455',
    website: '',
    summary: '',
    profiles: [
      {
        __typename: 'Profile',
        network: 'linkedin',
        username: 'aretecode',
        url: 'https://www.linkedin.com/in/james-wiens/',
      },
    ],
    address: '',
    postalCode: '',
    city: '',
    countryCode: '',
    region: '',
    resumeWebsite: '',
    skills: ['skill1'],
  },
  work: [
    {
      __typename: 'Work',
      company: 'Open Source',
      position: '',
      website: 'https://github.com/aretecode',
      startDate: '01/02/2013',
      endDate: 'current',
      summary: '',
      highlights: '',
      picture:
        'https://user-images.githubusercontent.com/4022631/55686780-04f0f980-5983-11e9-8152-204681b0840f.png',
    },
  ],
} as const
