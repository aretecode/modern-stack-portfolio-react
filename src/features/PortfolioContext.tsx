/**
 * can split PortfolioContext & PortfolioProvider
 */
import * as React from 'react'
import { Query } from 'react-apollo'
import { EMPTY_ARRAY, EMPTY_OBJ } from '../utils/EMPTY'
import { isObj } from '../utils/is'
import ResumeQuery from '../graphql/Resume'
import { defaultApolloStatePortfolio } from '../apolloState'
import { ResumeResponse, GraphqlProps, BasicsType, WorkType } from '../typings'

export interface PortfolioContextType {
  isLoading?: boolean
  basics: BasicsType
  work: WorkType[]
}

export const PortfolioContext = React.createContext<PortfolioContextType>(
  defaultApolloStatePortfolio
)

/**
 * @see /packages/graphql/src/graphql-modules/resume/utils.ts
 */
export function isValidResumeValue(
  response: GraphqlProps<ResumeResponse>
): response is GraphqlProps<Required<ResumeResponse>> {
  return (
    isObj(response) === true &&
    isObj(response.data) === true &&
    isObj(response.data!.resume) === true &&
    Array.isArray(response.data!.resume.work) &&
    isObj(response.data!.resume.basics) === true &&
    Object.keys(response.data!.resume.basics).length > 0
  )
}
/**
 * @todo @@perf can simplify & improve
 */
function fromResponseToSafeValue(response: GraphqlProps<ResumeResponse>) {
  const portfolio = isValidResumeValue(response)
    ? response.data!.resume
    : defaultApolloStatePortfolio

  const { refetch, loading } = response
  const value = { isLoading: !!loading, refetch, ...portfolio }
  return value
}

export class PortfolioProvider extends React.PureComponent {
  renderContext = (response: GraphqlProps<ResumeResponse>) => {
    const value = fromResponseToSafeValue(response)
    return (
      <PortfolioContext.Provider value={value}>
        {this.props.children}
      </PortfolioContext.Provider>
    )
  }

  render() {
    return (
      <Query<ResumeResponse>
        query={ResumeQuery}
        fetchPolicy="cache-and-network"
      >
        {this.renderContext}
      </Query>
    )
  }
}
