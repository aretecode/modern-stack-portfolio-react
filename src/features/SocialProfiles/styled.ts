import styled, { css } from 'styled-components'
import { MaterialIcon } from '../MaterialIcon'
import { StyledLink, StyledLinkProps } from '../Link'

export const StyledSocialProfileIcon = styled(MaterialIcon)`
  width: 2rem;
  height: 2rem;

  path:first-of-type {
    transition: fill 0.24s cubic-bezier(0.4, 0, 0.2, 1);
    fill: var(--theme-material-icon-fill);
  }
`

/**
 * @see https://material.io/design/motion/speed.html#easing
 * @todo when transitioning mobile to desktop, needs to morph
 * @note `order` is here because nothing is currently reused, would style in the page or larger component
 * @todo   transition: flex-direction 0.24s cubic-bezier(0, 0, 0.2, 1), width 0.24s cubic-bezier(0, 0, 0.2, 1);
 */
export const StyledSocialProfilesWrap = styled.aside`
  align-items: center;
  object-fit: contain;
  object-position: bottom;
  display: flex;
  max-width: 100%;
  background: transparent;

  @media (max-width: 1023px) {
    width: 75%;
    flex-direction: row;
    order: 2;
    height: unset;
    justify-content: unset;
    padding: 0 0 0.5rem 0;
  }
  @media (min-width: 1024px) {
    padding: 1rem;
    width: 75px;
    justify-content: space-around;
    flex-direction: column;
    flex-flow: column;
  }
`

/**
 * @todo HACK remove the twitter resize on mobile, make into overflow
 * @see https://github.com/aretecode/modern-stack-web-portfolio/issues/53
 *
 * @note 33% because parent is 75% and we have 3 items
 */
export const StyledProfileLink = styled(StyledLink)`
  @media (max-width: 1023px) {
    flex-basis: 33%;
    width: 33%;
    text-align: center;
    ${(props: StyledLinkProps) =>
      String(props.to).includes('twitter') &&
      css`
        display: none;
      `};
  }
`
