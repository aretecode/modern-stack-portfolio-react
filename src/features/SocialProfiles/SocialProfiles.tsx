import * as React from 'react'
import styled, { css } from 'styled-components'
import { ProfileType } from '../../typings'
import { MaterialIcon } from '../MaterialIcon'
import { PortfolioContext, PortfolioContextType } from '../PortfolioContext'
import { StyledLink } from '../Link'

/**
 * @todo material-ui timings
 * @todo when transitioning mobile to desktop, needs to morph
 *
 * @note `order` is here because nothing is currently reused, would style in the page or larger component
 */
export const StyledSocialProfilesWrap = styled.aside`
  padding: 1rem;
  width: 75px;
  display: inline-flex;
  justify-content: space-around;
  flex-direction: column;
  align-items: center;

  object-fit: contain;
  object-position: bottom;
  transition: flex-direction 0.24s ease-in-out, width 0.24s ease-in-out;

  @media (max-width: 1023px) {
    width: 75%;
    flex-direction: row;
    order: 2;
    height: unset;
    justify-content: unset;
    padding: 0 0 0.5rem 0;
  }
`

/**
 * @todo HACK remove the twitter resize on mobile, make into overflow
 * @see https://github.com/aretecode/modern-stack-web-portfolio/issues/53
 *
 * @note 33% because parent is 75% and we have 3 items
 */
export const StyledProfileLink = styled(StyledLink)`
  @media (max-width: 1023px) {
    flex-basis: 33%;
    width: 33%;
    text-align: center;
    ${(props: React.ComponentProps<typeof StyledLink>) =>
      String(props.to).includes('twitter') &&
      css`
        display: none;
      `};
  }
`

export const StyledSocialProfileIcon = styled(MaterialIcon)`
  width: 2rem;
  height: 2rem;
`

/**
 * I used a resize observer hook here
 * but it updated far too often
 * https://codesandbox.io/s/9yrwl14rp4
 *
 * @todo find or make a hook & store for resizing
 */
function ProfileItem(props: ProfileType) {
  const { url, username, network } = props
  return (
    <StyledProfileLink rel="me" to={url} title={username}>
      <StyledSocialProfileIcon icon={network} />
    </StyledProfileLink>
  )
}

export function renderProfileItem(profile: ProfileType, index?: number) {
  return <ProfileItem {...profile} key={profile.url} />
}

/**
 * @poc
 *    <img src="https://user-images.githubusercontent.com/4022631/55691849-01c82e80-59c0-11e9-8c5b-69f64a055b1f.png" />
 */
export default class SocialProfiles extends React.PureComponent {
  static contextType = PortfolioContext
  readonly context: PortfolioContextType

  render() {
    const { profiles, resumeWebsite } = this.context.basics

    return (
      <StyledSocialProfilesWrap {...this.props}>
        {Array.isArray(profiles) && profiles.map(renderProfileItem)}
        {resumeWebsite !== '' &&
          renderProfileItem({
            url: resumeWebsite,
            username: 'download resume',
            network: 'pdf',
          })}
      </StyledSocialProfilesWrap>
    )
  }
}
