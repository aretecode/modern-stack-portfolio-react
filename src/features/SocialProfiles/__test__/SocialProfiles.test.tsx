import * as React from 'react'
import { render } from '../../../../__tests__/render'
import { PortfolioContext } from '../../PortfolioContext'
import SocialProfiles from '../SocialProfiles'
import { defaultApolloStatePortfolio } from '../../../constants'

describe('SocialProfiles', () => {
  it('should render all for context', () => {
    const { container } = render(
      <PortfolioContext.Provider value={defaultApolloStatePortfolio as any}>
        <SocialProfiles />
      </PortfolioContext.Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
