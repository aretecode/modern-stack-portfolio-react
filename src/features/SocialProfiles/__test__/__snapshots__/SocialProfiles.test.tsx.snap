// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialProfiles should render all for context 1`] = `
.c1 {
  text-decoration: none;
  position: relative;
  letter-spacing: 0.2em;
  color: var(--color-link);
}

.c1:link,
.c1:visited {
  color: var(--color-link);
}

.c1:focus {
  outline: thin dotted;
}

.c1:hover {
  color: #aaa;
}

.c1:link {
  -webkit-tap-highlight-color: rgba(102,102,102,0.5);
}

.c0 {
  padding: 1rem;
  width: 75px;
  display: inline-flex;
  justify-content: space-around;
  flex-direction: column;
  align-items: center;
  object-fit: contain;
  object-position: bottom;
  transition: flex-direction 0.24s ease-in-out,width 0.24s ease-in-out;
}

.c3 {
  width: 2rem;
  height: 2rem;
}

@media (max-width:1023px) {
  .c0 {
    width: 75%;
    flex-direction: row;
    order: 2;
    height: unset;
    justify-content: unset;
    padding: 0 0 0.5rem 0;
  }
}

@media (max-width:1023px) {
  .c2 {
    flex-basis: 33%;
    width: 33%;
    text-align: center;
  }
}

<div>
  <aside
    class="c0"
  >
    <a
      class="c1 c2"
      href="https://www.linkedin.com/in/james-wiens/"
      rel="me"
      title="aretecode"
    >
      <svg
        class="c3"
        role="img"
        viewBox="0 0 24 24"
        xmlns="https://www.w3.org/2000/svg"
      >
        <title>
          linkedin icon
        </title>
        <path
          d="M16.893 20.452v-5.569c0-1.328-.027-3.037-1.852-3.037-1.853 0-2.136 1.445-2.136 2.939v5.667H9.351V9h3.414v1.561h.046c.477-.9 1.637-1.85 3.37-1.85 3.601 0 4.267 2.37 4.267 5.455v6.286h-3.555zM5.337 7.433a2.062 2.062 0 0 1-2.063-2.065 2.064 2.064 0 1 1 2.063 2.065zm1.782 13.019H3.555V9h3.564v11.452z"
        />
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
      </svg>
    </a>
  </aside>
</div>
`;
